tf2_gnn/models/graph_task_model.py

```diff
@@ -201,6 +196,28 @@ def _apply_gradients(
         if getattr(self, "_optimizer", None) is None:
             self._optimizer = self._make_optimizer()
 
+        # Filter out variables without gradients:
+        gradient_variable_pairs = [
+            (grad, var) for (grad, var) in gradient_variable_pairs if grad is not None
+        ]
+        clip_val = self._params.get("gradient_clip_value")
+        if clip_val is not None:
+            gradient_variable_pairs = [
+                (tf.clip_by_value(grad, -clip_val, clip_val), var)
+                for (grad, var) in gradient_variable_pairs
+            ]
+
+        clip_norm_val = self._params.get("gradient_clip_global_norm")
+        if clip_norm_val is not None:
+            grads = [grad for (grad, var) in gradient_variable_pairs]
+            clipped_grads = tf.clip_by_global_norm(grads, clip_norm=clip_norm_val)
+            gradient_variable_pairs = [
+                (clipped_grad, var)
+                for (clipped_grad, (_, var)) in zip(
```

Comment:
Wow, seriously awkward syntax, but there isn't an obvious way to make this better.

Reply:
Yeah, it's somewhat driven by the fact that clip_by_global_norm needs the full list of tensors (for the global norm), but then they need to be joined up again with the variable names... I thought about doing it in a more functional style (`zip(clip_by_global_norm(map fst ...), map snd)`) but that doesn't look any nicer.